include/nlohmann/detail/macro_scope.hpp

```diff
@@ -153,3 +153,14 @@
     basic_json<ObjectType, ArrayType, StringType, BooleanType,             \
     NumberIntegerType, NumberUnsignedType, NumberFloatType,                \
     AllocatorType, JSONSerializer>
+
+#ifndef JSON_USE_EXPLICIT_CONVERSIONS
+    #define JSON_EXPLICIT
+    #ifndef JSON_USE_IMPLICIT_CONVERSIONS
+        #warning Implicit conversions are being deprecated and will be \
```

Comment:
I realized that the `#warning` statement is not standard, and not supported on msvc. There is some solution like `#pragma message()` but it's not a real warning.

I think we could use `[[deprecated("...")]]` or the msvc non standard `__declspec(deprecated("..."))` for earlier versions in that case.

Reply:
I've added JSON_DEPRECATED_MSG, and tried building the tests by removing the `Wno-deprecated` warnings, it works \o/

Reply:
Thank you, the patch is great now. I would just wait for @nlohmann to do a review too and merge it!

Reply:
I like the idea of having a switch for the conversions, but I am still not convinced that implicit conversions should be deprecated.

Reply:
The main reason is because it is broken today. The pile of issues related to #958 is a good indicator of that. Not to mention the **why** is quite hard to grasp.

The required changes are quite minimal as well:

```cpp
std::string s;
s = j;
// becomes
std::string s;
j.get_to(s);

std::string s = j;
// becomes
std::string s(j);
```

Reply:
After thinking about it I think @nlohmann is right on not deprecate *right now* (I still think it should eventually). The compile switch must be there for a while before any deprecation, and the documentation should also rely on explicit conversion and it should recommend enabling the switch.

There is non-trivial cases that must be refactored when changing from implicit to explicit. Take this for example:
```c++
struct special_type {
    int a, b;
    std::string c, d;
};

special_type{value1, value2, value3, value4}; // works with implicit only
```

I really don't think a lot of users are relying on this, but it's still an annoyance. If we can recommend using the compiler switch for a while, warn users that the implicit operator will be deprecated in the future then the transition will be smooth.

The only way to recreate the behaviour above is to create a thin wrapper that allows casts:

```c++
struct implicit {
    json::value value;
    template<typename T, typename = decltype(std::declval<json::value>().get<T>())>
    operator T () {
        return value.get<T>();
    }
};

special_type{implicit{value1}, implicit{value2}, implicit{value3}, implicit{value4}};
```

I would like to see the implicit operator be deprecated too but I also agree that even though the change in the syntax is minimal, its use is widespread so it must be done very gradually.